Config flow for comfoconnect

beniwohli opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add config flow to ComfoConnect component. An existing YAML configuration is
imported. The selection of which sensors to add is done via options.

As this is my first contribution to Home Assistant, I'm not super familiar with
the inner workings, e.g. if it is correct to use `async_setup` to trigger the
import flow (I've seen a few other components do it this way). 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/19773

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

beniwohli committed:
implemented config flow for comfoconnect component

beniwohli committed:
Merge branch 'dev' into comfoconnect-config-flow

beniwohli committed:
use options for sensor selection, and add tests

beniwohli committed:
Merge branch 'dev' into comfoconnect-config-flow

beniwohli committed:
fix automatic import

beniwohli committed:
remove non-working unregister sensor code

homeassistant commented:
Hi @beniwohli, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#57002).

Once you do that we will be able to review and accept this pull request.

Thanks!

thecode commented:
>As this is my first contribution to Home Assistant, I'm not super familiar with
the inner workings, e.g. if it is correct to use async_setup to trigger the
import flow (I've seen a few other components do it this way).

Yes it is the way to do it

thecode commented on homeassistant/components/comfoconnect/__init__.py:
```diff
@@ -71,20 +96,49 @@ def setup(hass, config):
 
-    # Load platforms
-    discovery.load_platform(hass, "fan", DOMAIN, {}, config)
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _shutdown)
```
From looking at the `cbb.disconnect` method it looks like  it will not work if you call it more than once.
If the entry is unloaded and HA is stopped this will be called twice, it to prevent it wrap it with `entry.async_on_unload`

thecode commented on homeassistant/components/comfoconnect/__init__.py:
```diff
@@ -71,20 +96,49 @@ def setup(hass, config):
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _shutdown)
+    entry.async_on_unload(entry.add_update_listener(options_update_listener))
 
+    for platform in PLATFORM:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, platform)
+        )
```
```suggestion
    hass.config_entries.async_setup_platforms(entry, PLATFORM)
```
I also suggest to rename the const `PLATFORM` to `PLATFORMS` to make it more clear it is a list of platforms

thecode replied:
Maybe this was missed, please replace the loop with `hass.config_entries.async_setup_platforms(entry, PLATFORMS)` (it does exactly what is in your loop)

beniwohli replied:
ah, sorry, I misread your suggestion!

thecode commented on homeassistant/components/comfoconnect/const.py:
```diff
@@ -0,0 +1,13 @@
+from typing import Final
+
+DOMAIN: Final = "comfoconnect"
+PLATFORM: Final = ["sensor", "fan"]
```
```suggestion
PLATFORMS = ["sensor", "fan"]
```
Better to name it `PLATFORMS`, No need to mark all the consts with `Final`

beniwohli committed:
fix pylint issues

beniwohli committed:
implement code review suggestions from @thecode

beniwohli committed:
correctly apply suggestion from @thecode's review

beniwohli committed:
Merge branch 'dev' into comfoconnect-config-flow

michaelarnauts commented:
> @michaelarnauts thanks, great suggestions!
> 
> I'm not super happy that the user has to click on "Configure" after setting up the integration, to select sensors. I looked for a way to trigger the options dialog automatically after setup, but haven't found a solution. Maybe it would make sense to define a list of default sensors so the user sees at least something if they don't hit "Configure"?

I haven't looked into the config flows, but can't you have some wizard style with multiple steps? The 2nd step could be to select the sensors to subscribe for.

I also currently get an error untill I remove my yaml config. I'm not sure if this should happen.

thecode commented:
> I haven't looked into the config flows, but can't you have some wizard style with multiple steps? The 2nd step could be to select the sensors to subscribe for.

It is possible to have multiple steps, each config flow step is returning the next step, each step can have it's own form fields

> I also currently get an error untill I remove my yaml config. I'm not sure if this should happen.

Do you get an error or warning? A warning to remove the yaml config is OK, since once transitioned to a config flow the yaml is not needed and should be removed. 

Example for calling a step to pick a ZHA usb radio during user step:
https://github.com/home-assistant/core/blob/6a39119ccca329a536b867e62cccaea49eb503a1/homeassistant/components/zha/config_flow.py#L62

beniwohli committed:
discover bridges during config flow

also, make token and user agent non-configurable, as per suggestion from @michaelarnauts

beniwohli committed:
add back sensor selection to config flow, but store as options

this allows the user to add/remove sensors via options flow later on

beniwohli commented:
> But I guess we could store the result of `async_step_sensor_selection` in `options` instead of `data`, which would make it configurable in the options flow...

ce7aec2505 implements this approach.

beniwohli committed:
Merge branch 'dev' into comfoconnect-config-flow

michaelarnauts commented:
Looks awesome! Thanks! I just have some small string changes I would propose.

I'm also a bit worried about the error I still get when there still is config in the configuration.yaml file.

beniwohli committed:
Update homeassistant/components/comfoconnect/strings.json

Co-authored-by: Michaël Arnauts <michael.arnauts@gmail.com>

beniwohli committed:
Update homeassistant/components/comfoconnect/strings.json

Co-authored-by: Michaël Arnauts <michael.arnauts@gmail.com>

beniwohli committed:
Update homeassistant/components/comfoconnect/strings.json

Co-authored-by: Michaël Arnauts <michael.arnauts@gmail.com>

beniwohli committed:
update english translation file

beniwohli committed:
Merge branch 'dev' into comfoconnect-config-flow

beniwohli committed:
use DeviceInfo class for device_info attribute

beniwohli committed:
Merge branch 'dev' into comfoconnect-config-flow

beniwohli committed:
use STEP_ID constants

frenck commented on homeassistant/components/comfoconnect/__init__.py:
```diff
@@ -48,21 +53,41 @@
+) -> bool:
+    """Set up platform from a ConfigEntry."""
+    host = entry.data[CONF_HOST]
+    name = entry.data.get(CONF_NAME, DEFAULT_NAME)
+    token = entry.data.get(CONF_TOKEN, DEFAULT_TOKEN)
+    user_agent = entry.data.get(CONF_USER_AGENT, DEFAULT_USER_AGENT)
+    pin = entry.data.get(CONF_PIN, DEFAULT_PIN)
```
The defaults should already be in the entry data (added by the config flow).
There is no need to resolve them here?

frenck commented on homeassistant/components/comfoconnect/__init__.py:
```diff
@@ -102,15 +155,15 @@ def __init__(self, hass, bridge, name, token, friendly_name, pin):
 
-    def connect(self):
+    async def connect(self):
         """Connect with the bridge."""
         _LOGGER.debug("Connecting with bridge")
-        self.comfoconnect.connect(True)
+        await self.hass.async_add_executor_job(self.comfoconnect.connect, True)
```
Why make this an async method if you are passing it as an executor job anyways?
Just keep it a sync method in that case.

frenck commented on homeassistant/components/comfoconnect/__init__.py:
```diff
@@ -102,15 +155,15 @@ def __init__(self, hass, bridge, name, token, friendly_name, pin):
         """Disconnect from the bridge."""
         _LOGGER.debug("Disconnecting from bridge")
-        self.comfoconnect.disconnect()
+        await self.hass.async_add_executor_job(self.comfoconnect.disconnect)
```
Same as above

frenck commented on homeassistant/components/comfoconnect/config_flow.py:
```diff
@@ -0,0 +1,158 @@
+    ccb = ComfoConnectBridge(
+        hass,
+        bridge=bridges[0],
+        name=input_data.get(CONF_NAME, DEFAULT_NAME),
+        token=input_data.get(CONF_TOKEN, DEFAULT_TOKEN),
+        friendly_name=input_data.get(CONF_USER_AGENT, DEFAULT_USER_AGENT),
+        pin=input_data.get(CONF_PIN, DEFAULT_PIN),
```
These defaults should be the result of the config validation schema, and should not be needed here.

frenck commented on homeassistant/components/comfoconnect/config_flow.py:
```diff
@@ -80,14 +84,14 @@ async def async_step_user(self, user_input=None):
                 await self.async_set_unique_id(unique_id)
                 self._abort_if_unique_id_configured()
+                user_input[CONF_TOKEN] = secrets.token_hex(16)
+                user_input[CONF_USER_AGENT] = DEFAULT_USER_AGENT
```
CONF_USER_AGENT is hardcoded to the `DEFAULT_USER_AGENT` I suggest we remove it from the user_input and just use the default in places where needed.

frenck commented on homeassistant/components/comfoconnect/config_flow.py:
```diff
@@ -0,0 +1,158 @@
+
+    async def async_step_sensor_selection(self, user_input=None):
+        """Handle sensor selection step."""
+        schema = vol.Schema(
+            {
+                vol.Optional(
+                    CONF_SENSORS,
+                ): cv.multi_select({sensor.key: sensor.name for sensor in SENSOR_TYPES})
+            }
+        )
```
We should not offer this selection (especially not in the configuration flow). Instead, add them all and let the user decide.

michaelarnauts replied:
@frenck Adding them all used to give a performance impact on the bridge (although I haven't tested this recently), so it's best to only subscribe for the sensors you need. 

But maybe we should select a set of "minimal defaults"?

frenck replied:
> Adding them all used to give a performance impact on the bridge (although I haven't tested this recently), so it's best to only subscribe for the sensors you need.

In that case, it still should not be done this way. Instead, we can disable the less common sensors by default in the entity registry.

frenck commented on homeassistant/components/comfoconnect/config_flow.py:
```diff
@@ -0,0 +1,158 @@
+            errors=errors,
+        )
+
+    async def async_step_import(self, import_config):
+        """Handle a flow import."""
+        return await self.async_step_user(import_config)
```
Considering the defaults done in the validation, I think the data needs to be pre-processed here.

frenck commented on homeassistant/components/comfoconnect/config_flow.py:
```diff
@@ -0,0 +1,158 @@
+        if user_input is not None:
+            return self.async_create_entry(title="", data=user_input)
+
+        return self.async_show_form(
+            step_id=STEP_ID_INIT,
+            data_schema=vol.Schema(
+                {
+                    vol.Optional(
+                        CONF_SENSORS,
+                        default=self.config_entry.options.get(CONF_SENSORS),
+                    ): cv.multi_select(
+                        {sensor.key: sensor.name for sensor in SENSOR_TYPES}
+                    )
+                }
+            ),
```
Turning on/off entities/sensors is a Home Assistant default feature, we do not need a new construct/option flow for this.

frenck reviewed (changes requested):
Some initial comments, that have most likely impact on the rest of the implementation (so I've skipped tests and platform in this review for now).

michaelarnauts commented:
@beniwohli are you still interested in modifying this PR according to frenck's remarks? If not, I might want to look into this a bit later. I want it moved over to a config flow before the modifications for bypass control.

tnagels commented:
is anyone still actively working on this? Anything which can be done to finalize this code?

frank8the9tank commented:
I also like to know this, i really like to see this into HA and also the bypass control,
Just to add: in the latest update of HA then connection does not get restored when you connect with the zehnder app to the unit.
You get an error code in the HA log and the connection does not get restored,
Do you want me to open a bugreport for this or is this a known issue?

T0FFF commented:
@beniwohli thanks for this PR,
I'm also very interested by the config flow for this component to be able to use homeassistant.reload_config_entry.
Thanks all for your great job !

michaelarnauts commented:
@beniwohli Thanks a lot for your effort in this, but I'm going to close this PR, since it's implemented in my custom integration for now.

I still intend to submit this to Home Assistant core when it has seen more testing.
